src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1501,5 +1504,115 @@ private interface IWebSocketReceiveResultGetter<TResult>
             public WebSocketReceiveResult GetResult(int count, WebSocketMessageType messageType, bool endOfMessage, WebSocketCloseStatus? closeStatus, string closeDescription) =>
                 new WebSocketReceiveResult(count, messageType, endOfMessage, closeStatus, closeDescription);
         }
+
+        private class EnsureBuffer : IValueTaskSource
+        {
+            private readonly ManagedWebSocket _webSocket;
+            private readonly Action _onComplete;
+            private ManualResetValueTaskSourceCore<VoidResult> _mrvts;
+
+            private int _minimumRequiredBytes;
+            private CancellationToken _cancellationToken;
+            private bool _throwOnPrematureClosure;
+
+            private ValueTaskAwaiter<int> _awaiter;
+
+            public EnsureBuffer(ManagedWebSocket webSocket)
+            {
+                _webSocket = webSocket;
+                _onComplete = new Action(OnComplete);
+            }
+
+            public ValueTask EnsureBufferContainsAsync(int minimumRequiredBytes, CancellationToken cancellationToken, bool throwOnPrematureClosure)
+            {
+                _minimumRequiredBytes = minimumRequiredBytes;
+                _cancellationToken = cancellationToken;
+                _throwOnPrematureClosure = throwOnPrematureClosure;
+
+                EnsureBufferContains();
+
+                return new ValueTask(this, _mrvts.Version);
+            }
+
+            private void EnsureBufferContains()
+            {
+                // While we don't have enough data, read more.
+                while (_webSocket._receiveBufferCount < _minimumRequiredBytes)
+                {
+                    ValueTask<int> vt = ReadAsync();
+                    int numRead;
+                    if (vt.IsCompletedSuccessfully)
+                    {
+                        numRead = vt.Result;
+                    }
+                    else
+                    {
+                        _awaiter = vt.GetAwaiter();
+                        _awaiter.UnsafeOnCompleted(_onComplete);
+                        return;
+                    }
```

Comment:
Seems like the flow would be better here if you inverted the condition:
```C#
if (!vt.IsCompletedSuccessfully)
{
    _awaiter = vt.GetAwaiter().ConfigureAwait();
    _awaiter.UnsafeOnCompleted(_onComplete);
    return;
}

int numRead = vt.Result;
Debug.Assert(numRead >= 0, $"Expected non-negative bytes read, got {numRead}");
... // etc.
```

Reply:
Done